### ISSUE TITLE
State of internal CPU registers should be moved into a struct

### ISSUE THREAD

**Issue by JayFoxRox** [jarikomppa/emu8051#9]
I'm using emu8051 as part of another project and I require custom savestates.
However, I can't just save or load the entire struct from file because it contains pointers.

I have additional devices and ROM banks anyway, so I mostly care about the CPU register state.
Therefore these fields should be in a struct, or at least next to each other:

https://github.com/jarikomppa/emu8051/blob/b93d1da3a93b83175308e390526b476b1557977a/emu8051.h#L69-L70
https://github.com/jarikomppa/emu8051/blob/b93d1da3a93b83175308e390526b476b1557977a/emu8051.h#L79-L84

That way, one could easily copy them to a file (for now, I'll individually write each entry to a file - I guess the benefit is that it allows to handle endianess correctly).

**Comment by steveschnepp** [comment]
Can't you `memcpy` them ? 

As `int a[2]` is not a pointer, it is the real array directly inside the `struct`. 

Yes `emu.a` is a pointer, but it points *inside* the struct. And it is always reconstructed.

**Comment by JayFoxRox** [comment]
Yes, I can handle them perfectly fine as-is, but it would be much easier to separate the CPU register file from internal state of the emulator (such as callbacks).
This would simplify savestate handling.

Note how I currently cherry-pick registers for savestates: https://github.com/JayFoxRox/mtk-odd-emulator/blob/135043165130322730da7433cf04d639a7770dff/savestate.c#L23

I have to do that, because I also can't just dump the entire emu8051 struct, because it contains pointers

**Comment by steveschnepp** [comment]
> That way, one could easily copy them to a file (for now, I'll individually write each entry to a file - I guess the benefit is that it allows to handle endianess correctly).

Actually it has nothing to do with endianess. 

The CPU registers that are used are always located in the SFR. Those extra ones are only used for **assertion** with returning from an interrupt. And it is an array because it is indexed by the *interruption priority*. 

They are only read at https://github.com/jarikomppa/emu8051/blob/5dc681275151c4a5d7b85ec9ff4ceb1b25abd5a8/opcodes.c#L440-L446

I agree that `itr_psw` might have been less misleading than the current `int_psw` which makes a mental reference to `integer` instead of `interrupt`. That said, my ac29aa3380df3423895aba0b33893ea15482bb67 made it as least clear that it was, in fact, a `uint8_t` that is stored 😉 

So, you might even avoid to serialize them in the first place...
